tutorials/e-commerce-store-codelab/step-05-user-authentication.md

```diff
@@ -0,0 +1,381 @@
+  serverContext: any
+) {
+  const { getSession } = cookieSessionCreation(serverContext);
+  const session = await getSession(request.headers.get("Cookie"));
+  const userId = session.get("userId");
+  if (!userId) {
+    const searchParams = new URLSearchParams([["redirectTo", redirectTo]]);
+    redirect(`/account/login?${searchParams}`);
+  }
+  return userId;
+}
+```
+
+We can then import and call this function at the top of the loader functions of
+all the pages we want to guard against unauthorized access.
+
+```tsx
+import { requireUserId } from "~/lib/session.server";
+
+export async function loader({ request, context }: LoaderArgs): Promise<any> {
+  const db = buildDbClient(context);
+  const userId = await requireUserId(
+    { request, redirectTo: "/account/login" },
+    context
+  );
+  if (userId === undefined) {
+    return redirect("/account/login");
+  } else {
+    // The rest of the page's loader logic
+  }
+}
+```
+
+## The user dashboard and logging out
+
+We are going to add a user dashboard where at the very minimum the user's order
+history will be listed.
+
+Also, we need to be able to log the authenticated users out of our app, and we
+are going to implement this feature right in the dashboard page.
+
+Since we’ve set up Drizzle schema relations, getting the orders based on a user
+is a very easy task, all we need to do is perform a [findFirst] query on the users
+table listing the users orders in the returned data.
+
+Create an `accounts.dashboard.tsx` page inside `app/routes` and within the page’s
+loader, query and return the user’s details as follows.
+
+```tsx
+/* Page guard code */
+
+const user = await db.query.users.findFirst({
+  where: (users, { eq }) => eq(users.id, userId),
+  with: {
+    orders: true,
+  },
+});
+return { user };
+```
+
+We can then get this data in the dashboard page component and list the user’s
+orders as follows.
+
+```tsx
+const { user } = useLoaderData<typeof loader>();
+
+return (
+	<table>
+	  <thead>
+      <th>Date</th> th>Total Cost</th> <th>Shipping to</th>
+	  </thead>
+	  <tbody>
+	    {user.orders.map((order: any) => (
+	      <tr key={order.id}>
+	        <td>
+	          {formatDate(order.orders.createdAt)}
+	        </td>
+	        <td>
+	          {order.orders.finalAmount}$
+	        </td>
+	        <td>
+	          {order.orders.shippingAddress}
+	        </td>
+	      </tr>
+	    ))}
+	  </tbody>
+	</table>
+)
+```
+
+The `formatDate()` function called in the above code is simply a date formatting
+utility function exported from the `app/lib/utils.ts`.
+
+```ts
+// app/lib/utils.ts
+
+export function formatDate(date: number) {
+  const day = new Date(date * 1000).getDate();
+  const month = [
+    "Jan",
+    "Feb",
+    "Mar",
+    "Apr",
+    "May",
+    "Jun",
+    "Jul",
+    "Aug",
+    "Sep",
+    "Oct",
+    "Nov",
+    "Dec",
+  ][new Date(date * 1000).getMonth()];
+  const year = new Date(date * 1000).getFullYear();
+  return `${day} ${month}, ${year}`;
+}
```

Comment:
It's almost always better to respect the end user's locale for date formatting rather than implement them yourself. I suggest using an existing API that does this, such as [DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat) or some other library that does similar.